docs/specification/README.md

```diff
@@ -312,6 +312,53 @@ A Data End record indicates the end of the data section.
 | --- | --- | --- | --- |
 | 4 | data_section_crc | uint32 | CRC32 of all bytes from the beginning of the file up to the DataEnd record. A value of 0 indicates the CRC32 is not available. |
 
+### Custom Statistic (op=0x10)
+
+Records some statistic about the MCAP content not covered by the [Statistics](#statistics-op0x0b) record.
+
+| Bytes | Name           | Type    | Description                               |
+| ----- | -------------- | ------- | ----------------------------------------- |
+| 4 + N | name           | string  | the human-readable name of the statistic. |
+| 8     | value          | float64 | The statistic value.                      |
+| 4 + N | subject        | string  | The subject of this statistic.            |
+| 4 + N | display_format | string  | how this statistic should be displayed.   |
+
+#### Custom statistic subjects
+
+Custom Statistics specify a subject, which indicates what this statistic is about. Some subjects are well-known:
+
+| Subject | Description |
+| --- | --- |
+| `messages` | a statistics about all messages |
+| `channels` | a statistic about all channels |
+| `schemas` | a statistic about all schemas |
+| `metadatas` | a statistic about all metadata records |
+| `attachments` | a statistic about all attachment records |
+| `chunks` | a statistic about all chunks |
+| `this_channel` | a statistic about the channel record immediately preceding |
```

Comment:
Few thoughts on this -
1. I do not like that this introduces a new naming scheme with "this_*". I think we can avoid that.

2. I don't get how these will get recorded. I would expect the custom statistics records to get recorded into a custom statistics portion of the summary section (i.e these are not intermixed with the other portions of the summary section - there is a summary offset record for custom statistic, pointing at a contiguous run of these).

So what's the "immediately preceding" channel? Likewise for metadata/attachment/chunk.

3. Is it necessary to have statistics for both "this channel" and "channels"? If you are recording the "this channel" stats, would you be able to synthesize what you want for "channels"? And would tooling be able to do generic handling based on types (e.g for numeric types, compute the averages, distributions, etc maybe) to do a good enough job of covering everything?

If we can do that, we avoid forcing users to record both "this_chunk: message_count" and "chunks: average message count".

Reply:
Couple alternatives to throw out -

Single record: downside - "id" and opcode are abused.
```
Custom Statistic:
name string
value float64
op OpCode
id uint64

// Legal OpCodes are Channel and Chunk, as well as 0x00 which denotes a full-file statistic. 
```

Separate records
```
FileStatistic
name string
value float64

ChannelStatistic
name string
value float64
channelID uint16

ChunkStatistic
name string
value float64
offset uint64

```

In either case, we'd use the existing index structure. So for each record type getting introduced, there would be a summary offset record with that ID, pointing up at a contiguous run of those records in the summary section.

we can include display format on these as you have - haven't thought that through but seems like a reasonable idea to me.

Reply:
> If you are recording the "this channel" stats, would you be able to synthesize what you want for "channels"?

I considered allowing the Custom Specification record to specify how the reader should aggregate when more than one is concerned - would be neccessary for making "all channels" stats from individual stats as you say. It would also be neccessary to preserve custom stats across merges. The problem is that there are several ways for this to work (weighted average, min, max, sum, etc.) which is hard to specify ahead of time. i also considered embedding a little DSL to do this maths (something like `bc` notation or APL). In the end I don't think it's worth the API surface and implementation burden for the reader.